src/modules/analytics/withAnalytics.js

```diff
@@ -1,141 +1,208 @@
-      if (!bindEvents) {
-        return this.props;
+
+      // Consumers can tell us that certain analytics events are inherently bound
+      // to a callback prop. This means that whenever the event is raised we
+      // should capture it and pass it as an extra argument to that callback prop,
+      // rather than letting the consumer handle the event as normal.
+      bindEventsToProps = (props: Props): Props => {
+        const bindEventsMap = props.bindEventsToProps;
+        if (!bindEventsMap) {
+          return props;
+        }
+
+        const boundPropCallbacks = Object.keys(bindEventsMap)
+          .reduce((bound, eventName) => {
+            const propName = bindEventsMap[eventName];
+            const providedCallback = props[propName];
+            if (!providedCallback) {
+              return bound;
+            }
+            const boundCallback = (...args) => {
+              const event = this.getEnqueuedEvent(eventName);
+              providedCallback(...args, event);
+            };
+            return {
+              ...bound,
+              [propName]: boundCallback,
+            };
+          }, {});
+
+        return (({ ...props, ...boundPropCallbacks }: any): Props);
       }
 
-      return Object.keys(bindEvents)
-        .reduce((bound, eventName) => {
-          const propName = bindEvents[eventName];
-          const providedCallback = this.props[propName];
-          if (!providedCallback) {
-            return bound;
-          }
-          const boundCallback = (...args) => {
-            const event = this.getEnqueuedEvent(eventName);
-            providedCallback(...args, event);
-          };
-          return {
-            ...bound,
-            [propName]: boundCallback,
-          };
-        }, {});
-    }
-
-    render() {
-      const {
-        analyticsNamespace,
-        ...permittedProps
-      } = this.props;
-
-      const props = {
-        ...permittedProps,
-        ...this.bindEventsToProps(),
-        createAnalyticsEvent: this.createAnalyticsEvent,
-      };
+      // Consumers can provide a map of callbacks to event creators to allow creating events
+      // within certain components.
+        mapCreateEventsToProps = (props: Props): Props => {
+        if (!createEventMap) {
+          return props;
+        }
+
+        const modifiedPropCallbacks = Object.keys(createEventMap)
+          .reduce((modified, propCallbackName) => {
+            const eventCreator = createEventMap[propCallbackName];
+            const providedCallback = props[propCallbackName];
+            if (!providedCallback) {
+              return modified;
+            }
+            const modifiedCallback = (...args) => {
+              if (typeof eventCreator === 'string') {
+                this.createAnalyticsEvent(eventCreator).raise();
+              } else if (typeof eventCreator === 'function') {
+                eventCreator(this.createAnalyticsEvent, props).raise();
```

Comment:
I would suggest that we allow them to return `null` from this function (so we'll have do a null check before we call `.raise`). It's probably an edge case, but they might want to fire an event directly from the callback prop without ever raising an event.